tools/analyzer_plugin/lib/src/diagnostic/create_ref_usage.dart

```diff
@@ -0,0 +1,96 @@
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/visitor.dart';
+import 'package:analyzer_plugin/protocol/protocol_common.dart';
+import 'package:over_react_analyzer_plugin/src/diagnostic_contributor.dart';
+import 'package:over_react_analyzer_plugin/src/util/function_components.dart';
+import 'package:over_react_analyzer_plugin/src/util/react_types.dart';
+
+const _desc = r'Prefer useRef over createRef within a function component.';
+// <editor-fold desc="Documentation Details">
+const _details = r'''
+
+This diagnostic detects when `createRef`, instead of `useRef`, is used inside a function component.
+
+**BAD:** 
+```
+import 'package:over_react/over_react.dart';
+
+part 'create_ref_usage.over_react.g.dart';
+
+UiFactory<FooProps> Foo = uiFunction(
+  (props) {
+    final ref = createRef();
+
+    return (Dom.div()..ref = ref)();
+  },
+  _$FooConfig, // ignore: undefined_identifier
+);
+
+mixin FooProps on UiProps {}
+```
+
+**GOOD:**
+```
+import 'package:over_react/over_react.dart';
+
+part 'create_ref_usage.over_react.g.dart';
+
+UiFactory<FooProps> Foo = uiFunction(
+  (props) {
+    final ref = useRef();
+
+    return (Dom.div()..ref = ref)();
+  },
+  _$FooConfig, // ignore: undefined_identifier
+);
+
+mixin FooProps on UiProps {}
+```
+
+''';
+// </editor-fold>
+
+class CreateRefUsageDiagnostic extends DiagnosticContributor {
+  @DocsMeta(_desc, details: _details)
+  static const code = DiagnosticCode(
+    'over_react_create_ref_usage',
+    "Prefer useRef over createRef within a function component.",
+    AnalysisErrorSeverity.WARNING,
+    AnalysisErrorType.STATIC_WARNING,
+  );
+
+  static final fixKind = FixKind(code.name, 400, "Replace with useRef");
+
+  @override
+  computeErrors(result, collector) async {
+    result.unit!.accept(FunctionComponentVisitor((component) async {
```

Comment:
I just noticed that this visitor body is `async`, but the result of those futures isn't awaited. Can we change this to `getAllFunctionComponents`, so we can iterate them in a for-loop or something where we can `await` calls properly?

Reply:
Oh I just thought, we also should warn when people use `createRef` in custom hook functions. I'm down to do that in a follow up PR, though. It should just be iterating over `getAllHookUsages` as well.